internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -5,61 +5,84 @@
 package asset
 
 import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
 	"fmt"
 	"io"
 	"io/fs"
 	"path/filepath"
 
+	"github.com/aws/copilot-cli/internal/pkg/manifest"
 	"github.com/spf13/afero"
+	"golang.org/x/sync/errgroup"
 )
 
-// UploadFunc is the function signature to upload contents to a destination.
-type UploadFunc func(dest string, contents io.Reader) (url string, err error)
+// CacheUploader uploads local asset files.
+type CacheUploader struct {
+	// FS is the file system to use.
+	FS afero.Fs
 
-// UploadOpts contains optional configuration for uploading assets.
-type UploadOpts struct {
-	Reincludes []string   // Relative path under source to reinclude files that are excluded in the upload.
-	Excludes   []string   // Relative path under source to exclude in the upload.
-	UploadFn   UploadFunc // Custom implementation on how to upload the contents under a file. Defaults to S3UploadFn.
-	Recursive  bool       // Whether to walk recursively.
+	// Upload is the function called when uploading a file.
+	Upload func(path string, contents io.Reader) (string, error)
+
+	// PathPrefix is the path to prefix any hashed files when uploading to the cache.
+	PathPrefix string
+
+	// AssetMappingPath is the path the upload the asset mapping file to.
+	AssetMappingPath string
 }
 
-// Upload uploads static assets to Cloud Storage and returns uploaded file URLs.
-func Upload(fs afero.Fs, source, destination string, opts *UploadOpts) ([]string, error) {
-	matcher := buildCompositeMatchers(buildReincludeMatchers(opts.Reincludes), buildExcludeMatchers(opts.Excludes))
-	var urls []string
-	info, err := fs.Stat(source)
-	if err != nil {
-		return nil, fmt.Errorf("get stat for file %q: %w", source, err)
-	}
-	paths := []string{source}
-	if info.IsDir() {
-		files, err := afero.ReadDir(fs, source)
-		if err != nil {
-			return nil, fmt.Errorf("read directory %q: %w", source, err)
-		}
-		paths = make([]string, len(files))
-		for i, f := range files {
-			paths[i] = filepath.Join(source, f.Name())
+type cached struct {
+	localPath string
+	content   io.Reader
+
+	Path            string `json:"path"`
+	DestinationPath string `json:"dest_path"`
```

Comment:
nit: consider using camelCase (https://stackoverflow.com/a/19287394/1201381)
```suggestion
	DestinationPath string `json:"destPath"`
```